clang/lib/Sema/SemaDecl.cpp

```diff
@@ -11145,7 +11148,24 @@ bool Sema::areMultiversionVariantFunctionsCompatible(
     FunctionType::ExtInfo OldTypeInfo = OldType->getExtInfo();
     FunctionType::ExtInfo NewTypeInfo = NewType->getExtInfo();
 
-    if (OldTypeInfo.getCC() != NewTypeInfo.getCC())
+    const auto *OldFPT = OldFD->getType()->getAs<FunctionProtoType>();
+    const auto *NewFPT = NewFD->getType()->getAs<FunctionProtoType>();
+
+    bool ArmStreamingCCMismatched = false;
+    // Locally streaming does not affect the calling convention.
+    if (OldFPT && NewFPT && !OldFD->hasAttr<ArmLocallyStreamingAttr>() &&
+        !NewFD->hasAttr<ArmLocallyStreamingAttr>()) {
```

Comment:
Why is it necessary to check for `ArmLocallyStreamingAttr` here?

Reply:
Because arm_locally_streaming is compatible with everything else, therefore if one of the two declarations has it, then we should not check whether the other declaration has arm_streaming or arm_streaming_compatible.

Reply:
Are you saying that `ok_arm_streaming` and `ok_arm_streaming_compatible` from my example in `clang/test/Sema/aarch64-fmv-streaming.c` should report an error? If that's not what you meant, then we must not remove this check. Doing so will diagnose a mismatch because __arm_locally_streaming does not set neither of `FunctionType::SME_PStateSMEnabledMask` or `FunctionType::SME_PStateSMCompatibleMask`.

Reply:
Perhaps I am mistaken. Do we need to add arm_streaming/arm_streaming_compatible to an arm_locally_streaming callee if the other versions have it?

Reply:
> Are you saying that ok_arm_streaming and ok_arm_streaming_compatible from my example in clang/test/Sema/aarch64-fmv-streaming.c should report an error? 

Yes, that's right, that is because for:
```
(A) __attribute__((target_clones("sve", "simd"))) void ok_arm_streaming(void) __arm_streaming {}
(B) __arm_locally_streaming __attribute__((target_version("sme2"))) void ok_arm_streaming(void) {}
```
(A) needs to be called in streaming mode, whereas (B) should be called in non-streaming mode.

I also think that the following case should give a diagnostic:
```
__attribute__((target_clones("sve", "simd"))) void bad_mixed_streaming2(void) {}

// expected-error@+1 {{multiversioned function declaration has a different calling convention}}
__arm_locally_streaming __attribute__((target_version("sme2"))) void bad_mixed_streaming2(void) __arm_streaming {}
```

The check for `__arm_locally_streaming` prevents that at the moment.

Reply:
I also think there is no need to disallow cases where one of the functions is `__arm_streaming_compatible`, e.g.

```
__attribute__((target_clones("sve", "simd"))) void n_callee(void) {}
__attribute__((target_version("sme2"))) void n_callee(void) __arm_streaming_compatible {}
__attribute__((target_version("default"))) void n_callee(void) {}

void s_caller(void) __arm_streaming {
  n_callee();
}
```
because all three versions of `n_callee` can be called from non-streaming mode, so from the perspective of the implementation of `s_caller`, it's fine for it to just do:
```
smstop
bl n_callee
smstart
```

Reply:
Thanks for clarifying! Ok, I'll fix the interaction with locally_streaming in a new revision. Your last example cannot be supported for the reason I explain on the commit message.